### PR TITLE
Add device vjp support for jax (non-jit) gradients

### PR DESCRIPTION
**Pull request by albi3ro**
[sc-39465]

This PR ports jax (non jit) over to the jacobian product calculator framework and thus adds support for device vjps.

Caveats:
There is only device vjp support for `jax.grad` and `jax.vjp`. `jax.jacobian` is calculated using a batch dimension on `dy`, and `adjoint_vjp` does not yet support a batch dimension on the `dy`.

Test Failure:
For some reason, jax is trying to unflatten a `QubitUnitary` with a raw python object as the matrix. I have no idea why it is trying to do that.

### CODE REVIEW

File: pennylane/tape/qscript.py [hunk 1]
```diff
@@ -765,7 +765,7 @@ def bind_new_parameters(self, params: Sequence[TensorLike], indices: Sequence[in
 
         new_tape = self.__class__(new_operations, new_measurements, shots=self.shots)
         new_tape.trainable_params = self.trainable_params
-
+        print(new_tape.circuit)
```
Comment (albi3ro) [inline pennylane/tape/qscript.py:768]:
```suggestion
```

**Comment by albi3ro** [comment]
closed in favor of #4878

**Comment by albi3ro** [comment]
> @albi3ro do you need to keep the branch open?

I'd like to keep it here until I merge the replacement branch.
